File: python/graphs/1020-number-of-enclaves.py
from typing import List


def numEnclaves(grid: List[List[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    
    stack = []
    for i in range(rows):
        if grid[i][0]: stack.append((i, 0))
        if grid[i][cols - 1]: stack.append((i, cols - 1))
    
    for j in range(cols):
        if grid[0][j]: stack.append((0, j))
        if grid[rows - 1][j]: stack.append((rows - 1, j))
    
    while stack: # dfs 
        i, j = stack.pop()
        grid[i][j] = 0 # mark as seen 
        for ii, jj in (i - 1, j), (i, j - 1), (i, j + 1), (i + 1, j): 
            if 0 <= ii < rows and 0 <= jj < cols and grid[ii][jj]: stack.append((ii, jj))
    
    return sum(map(sum, grid))

File: cpp/graphs/1020-number-of-enclaves.cpp
#include <stack>
#include <utility>
#include <vector>

using namespace std;


int numEnclaves(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();
    
    stack<pair<int, int>> s;
    
    for (int i = 0; i < m; i++) {
        if (grid[i][0]) s.push({i, 0});
        if (grid[i][n-1]) s.push({i, n-1});
    }
    
    for (int i = 0; i < n; i++) {
        if (grid[0][i]) s.push({0, i});
        if (grid[m-1][i]) s.push({m-1, i});
    }
    
    while (!s.empty()) {
        pair<int, int> cur = s.top();
        s.pop();
        
        int i = cur.first, j = cur.second;
        grid[i][j] = 0;
        
        for (auto neighbor : vector<vector<int>>({{i - 1, j}, {i + 1, j}, {i, j - 1}, {i, j + 1}})) {
            int ii = neighbor[0], jj = neighbor[1];
            if (ii >= 0 && ii < m && jj >= 0 && jj < n && grid[ii][jj]) s.push({ii, jj});
        }
    }
    
    int sum = 0;
    for (auto row : grid) {
        for (auto col : row) {
            sum += col;
        }
    }
    
    return sum;
}